packages/gatsby-plugin-sharp/src/image-data.ts

```diff
@@ -0,0 +1,230 @@
+
+  metadata = { width, height, density, format, dominantColor }
+  metadataCache.set(file.internal.contentDigest, metadata)
+  return metadata
+}
+
+export interface IImageDataProps {
+  file: FileNode
+  args: ISharpGatsbyImageArgs
+  cache: GatsbyCache
+  reporter: Reporter
+}
+
+export interface IImageDataArgs {
+  file: FileNode
+  args: ISharpGatsbyImageArgs
+  pathPrefix: string
+  cache: GatsbyCache
+  reporter: Reporter
+}
+
+export async function generateImageData({
+  file,
+  args,
+  pathPrefix,
+  reporter,
+  cache,
+}: IImageDataArgs): Promise<ISharpGatsbyImageData | undefined> {
+  const {
+    layout = `fixed`,
+    placeholder = `dominantColor`,
+    tracedSVGOptions = {},
+  } = args
+
+  const metadata = await getImageMetadata(file, placeholder === `dominantColor`)
+
+  const imageSizes: {
+    sizes: Array<number>
+    presentationWidth: number
+    presentationHeight: number
+    aspectRatio: number
+    isTopSizeOverriden: boolean
+  } = calculateImageSizes({
+    file,
+    layout,
+    ...args,
+    imgDimensions: metadata,
+    reporter,
+  })
+
+  const transforms = imageSizes.sizes.map(outputWidth => {
+    const width = Math.round(outputWidth)
+    const transform = createTransformObject({
+      ...args,
+      width,
+      height: Math.round(width / imageSizes.aspectRatio),
+      toFormat: args.toFormat || metadata.format,
+    })
+
+    if (pathPrefix) transform.pathPrefix = pathPrefix
+    return transform
+  })
+
+  const images = batchQueueImageResizing({
+    file,
+    transforms,
+    reporter,
+  })
+
+  const srcSet = getSrcSet(images)
+
+  const widthOfMaxSize = imageSizes.isTopSizeOverriden
+    ? metadata.width
+    : imageSizes.presentationWidth
+
+  const sizes = args.sizes || getSizes(widthOfMaxSize, layout)
+
+  const primaryIndex = imageSizes.sizes.findIndex(
+    size => size === widthOfMaxSize
+  )
+
+  if (primaryIndex === -1) {
+    reporter.panic(
+      `No image of the specified size found. Images may not have been processed correctly.`
+    )
+  }
+
+  const primaryImage = images[primaryIndex]
+
+  if (!images?.length) {
+    return undefined
+  }
+  const imageProps: ISharpGatsbyImageData = {
+    layout,
+    placeholder: undefined,
+    images: {
+      fallback: {
+        src: primaryImage.src,
+        srcSet,
+        sizes,
+      },
+      sources: [],
+    },
+  }
+
+  if (args.webP) {
+    const transforms = imageSizes.sizes.map(outputWidth => {
+      const width = Math.round(outputWidth)
+      const transform = createTransformObject({
+        ...args,
+        width,
+        height: Math.round(width / imageSizes.aspectRatio),
+        toFormat: `webp`,
+      })
+      return transform
+    })
+
+    const webpImages = batchQueueImageResizing({
+      file,
+      transforms,
+      reporter,
+    })
+    const webpSrcSet = getSrcSet(webpImages)
+
+    imageProps.images.sources?.push({
+      srcSet: webpSrcSet,
+      type: `image/webp`,
+      sizes,
+    })
```

Comment:
Follow up:
Any reason why we don't append webp to the previous batch? Wouldn't that speed up things? (less active jobs)

Or isn't it possible?

Reply:
Yeah, it should be possible. I'm goign to handle it differently with the updated API